juju/juju
# refactor: use model config service in caasmodeloperator facade

Refactor the caasmodeloperator facade to use a model config service instead of state. The facade uses the new `ModelConfig` and `Watch` methods instead of the state methods.

Unit tests have been changed accordingly, with the ModelConfigService mocked out by a gomock-generated mock.

As a result, many of the methods/interfaces/structs in `state.go` were able to be removed.

## Checklist




## QA steps

We are going to force a controller upgrade and check that the model operator pod is updated.

First, make sure you have an account on [Docker Hub](https://hub.docker.com/) and that you have `docker login`-ed on your local machine.

Build Juju and set up the image repository:
```bash
export JUJU_BUILD_NUMBER=0
export DOCKER_USERNAME=docker.io/<your-docker-username>
make seed-repository
make install   # or go-build
make simplestreams push-release-operator-image
```

In a new terminal window, start the simplestreams server.
```
$ cd _build/simplestreams/tools
$ python3 -m http.server
Serving HTTP on 0.0.0.0 port 8000 (http://0.0.0.0:8000/) ...
```

Back in the original terminal, we can bootstrap Juju with the right simplestreams & repository (providing your computer's IP address as the `agent-metadata-url`):
```
$ juju bootstrap minikube --config agent-metadata-url=http://192.168.1.25:8000/ --config caas-image-repo=$DOCKER_USERNAME
$ juju switch controller
$ juju status
Model       Controller  Cloud/Region  Version    Timestamp
controller  minikube    minikube      4.0-beta4  14:03:34+10:00
```

Go to `version/version.go` and increment the version to `4.0-beta5`. Now rebuild Juju and update the simplestreams/image repo:
```
make install   # or go-build
make simplestreams push-release-operator-image
```

Now we should be able to upgrade the controller.
```
$ juju upgrade-controller
best version:
    4.0-beta5
started upgrade to 4.0-beta5
$ juju status
Model       Controller  Cloud/Region  Version    Timestamp
controller  minikube    minikube      4.0-beta5  14:06:12+10:00
$ kubectl -ncontroller-minikube describe deployment modeloperator | grep jujud-operator:
    Image:      docker.io/hpidcock/jujud-operator:4.0-beta5
```

## Links

**Jira card:** JUJU-6100
